arrow/obsidian.c

```diff
@@ -0,0 +1,279 @@
+      // `interval` from an exponential distribution.
+      sample = (double) rand() / RAND_MAX;
+      interval = -log(1 - sample) / total;
+      point = ((double) rand() / RAND_MAX) * total;
+
+      // Based on the random sample, find the event that it corresponds to by
+      // iterating through the propensities until we surpass our sampled value
+      choice = 0;
+      progress = 0.0;
+      while(progress + propensities[choice] < point) {
+        progress += propensities[choice];
+        choice += 1;
+      }
+
+      // If we have surpassed the provided duration we can exit now
+      if (choice == -1 || (now + interval) > duration) {
+        break;
+      }
+
+      // Increase time by the interval sampled above
+      now += interval;
+
+      // Record the information about the chosen event this step
+      time[step] = now;
+      events[step] = choice;
+
+      // For each substrate involved in this reaction, update the ongoing state
+      // with its value from the stoichiometric matrix
+      substrates_length = substrates_lengths[choice];
+      for (involve = 0; involve < substrates_length; involve++) {
+        index = substrates_indexes[choice] + involve;
+        adjustment = stoichiometry[choice * substrates_count + substrates[index]];
+        outcome[substrates[index]] += adjustment;
+      }
+
+      // Find which propensities depend on this reaction and therefore need to be
+      // updated in the next step (the rest of the propensities will not change)
+      update_length = dependencies_lengths[choice];
+      for (up = 0; up < update_length; up++) {
+        index = dependencies_indexes[choice] + up;
+        update[up] = dependencies[index];
+      }
+
+      // Advance
+      step += 1;
+
+      // If our step has advanced beyond the current `event_bounds`, double the
+      // `event_bounds` and reallocate these arrays with the new size
+      if (step >= event_bounds) {
+        double * new_time = malloc((sizeof (double *)) * event_bounds * 2);
+        memcpy(new_time, time, (sizeof (double *)) * event_bounds);
+        free(time);
+        time = new_time;
+
+        long * new_events = malloc((sizeof (long *)) * event_bounds * 2);
+        memcpy(new_events, events, (sizeof (long *)) * event_bounds);
+        free(events);
+        events = new_events;
+
+        event_bounds *= 2;
+      }
+    }
+  }
+
+  // Construct the `evolve_result` from the results of performing steps until either
+  // the desired duration was achieved or there are no more reactions to perform.
+  // We return:
+  //   * How many steps were performed
+  //   * An array of the time of each event
+  //   * An array of each event that occurred each time step
+  //   * The resulting state of the system after the chosen reactions were applied
+  evolve_result result = {step, time, events, outcome};
+
+  // Clean up the transient allocations
+  free(propensities);
+  free(update);
```

Comment:
I know little about C and even less about connecting C with Python.  I'm sure it's a good practice but I'm curious about the necessity of these `free()` calls.  Would we leak memory without these?

Reply:
To be fair, I was basing not freeing `time`, `events` and `outcome` on this documentation: https://docs.scipy.org/doc/numpy-1.15.0/reference/c-api.array.html#c.PyArray_NewFromDescr especially the big red "Warning" box that says not to deallocate pointers that are getting used to construct numpy arrays for passing back into python. Also it looks like the value NPY_OWNDATA is set which is still potentially unclear but to me reads that numpy takes over responsibility of calling `free` on these pointer. I may be wrong however and this should be verified. The other sign is that I did not observe the memory for the simulation growing as symptomatic of a memory leak though I have not done an official measurement on this. I'll try that soon and report back.

If I do need to call free on these pointers then I need to know when numpy is done with them, which seems difficult to achieve. Probably they would have to be wrapped in an object with an explicit deallocate method. Simpler perhaps now that I think of it would be to construct an empty numpy array from C and fill it with the values from those three arrays and then deallocate them. That will solve the problem if it is shown to be leaking, I'll investigate further.

Reply:
Ah. Memory management is subtle! So if numpy takes ownership, then it just needs a comment. I like the approach of constructing a numpy array.